airbyte-workers/src/main/java/io/airbyte/workers/internal/StateMetricsTracker.java

```diff
@@ -0,0 +1,241 @@
+  private static final int EPOCH_TIME_SIZE = Long.BYTES;
+  private static final int BYTE_ARRAY_SIZE = STATE_HASH_SIZE + EPOCH_TIME_SIZE;
+
+  private final List<byte[]> stateHashesAndTimestamps;
+  private final Map<String, List<byte[]>> streamStateHashesAndTimestamps;
+  private LocalDateTime firstRecordReceivedAt;
+  private final LocalDateTime lastStateMessageReceivedAt;
+  private Long maxSecondsToReceiveSourceStateMessage;
+  private Long meanSecondsToReceiveSourceStateMessage;
+  private Long maxSecondsBetweenStateMessageEmittedandCommitted;
+  private Long meanSecondsBetweenStateMessageEmittedandCommitted;
+  private final AtomicLong totalSourceEmittedStateMessages;
+  private final AtomicLong totalDestinationEmittedStateMessages;
+  private long remainingCapacity;
+  private Boolean capacityExceeded;
+
+  public StateMetricsTracker(final Long memoryLimitBytes) {
+    this.stateHashesAndTimestamps = new ArrayList<>();
+    this.streamStateHashesAndTimestamps = new HashMap<>();
+    this.firstRecordReceivedAt = null;
+    this.lastStateMessageReceivedAt = null;
+    this.maxSecondsToReceiveSourceStateMessage = 0L;
+    this.meanSecondsToReceiveSourceStateMessage = 0L;
+    this.maxSecondsBetweenStateMessageEmittedandCommitted = 0L;
+    this.meanSecondsBetweenStateMessageEmittedandCommitted = 0L;
+    this.totalSourceEmittedStateMessages = new AtomicLong(0L);
+    this.totalDestinationEmittedStateMessages = new AtomicLong(0L);
+    this.remainingCapacity = memoryLimitBytes;
+    this.capacityExceeded = false;
+  }
+
+  public synchronized void addState(final AirbyteStateMessage stateMessage, final int stateHash, final LocalDateTime timeEmitted)
+      throws StateMetricsTrackerException {
+    final long epochTime = timeEmitted.toEpochSecond(ZoneOffset.UTC);
+
+    if (capacityExceeded || remainingCapacity < BYTE_ARRAY_SIZE) {
+      capacityExceeded = true;
+      throw new StateMetricsTrackerException("Memory capacity is exceeded for StateMetricsTracker.");
+    }
+
+    if (AirbyteStateType.STREAM == stateMessage.getType()) {
+      addStateMessageToStreamToStateHashTimestampTracker(stateMessage, stateHash, epochTime);
+    } else {
+      // do not track state message timestamps per stream for GLOBAL or LEGACY state
+      final byte[] stateTimestampByteArray = populateStateTimestampByteArray(stateHash, epochTime);
+      stateHashesAndTimestamps.add(stateTimestampByteArray);
+      remainingCapacity -= stateTimestampByteArray.length;
+    }
+  }
+
+  public synchronized void updateStates(final AirbyteStateMessage stateMessage, final int stateHash, final LocalDateTime timeCommitted) {
+    final LocalDateTime startingTime;
+    if (AirbyteStateType.STREAM == stateMessage.getType()) {
+      final StreamDescriptor streamDescriptor = stateMessage.getStream().getStreamDescriptor();
+      final String streamNameAndNamespace = streamDescriptor.getName() + streamDescriptor.getNamespace();
```

Comment:
namespace can be null, but we need to include it since stream descriptors can have the same name but different namespaces. I'm ok with no separator and leaving it as 'null' in the key since this isn't really being read by humans, but I can change it (if that's what you're saying?)

Reply:
I was thinking in terms of NullPointerException and collisions more than readability.

Reply:
I was thinking in terms of NullPointerException and collisions more than readability.

Reply:
oh you mean collisions in case there's one stream with name: 'abc', namespace: 'def' and another with name: 'ab' and namespace: 'cdef' - yes theoretically possible, separator makes sense. I think we're ok with null pointer exceptions though.